args<-commandArgs(TRUE)
#print(args)
TFile<-tempfile()
#TFile
sink(TFile)

require(R2HTML)


# 1 = Parameter will be Path
# 2 = name of GenoFile
# 3 = Phenotypic File
# 4 = Result File Name
# 5 = Missing %
# 6 = MAF 
# 7=  PIC = 0
# 8 = Method Name
# 9 = No of Core
# 10 = Missing CH
# 11 = ridgeRegression
# 12 = kinshipGauss
# 13 = BayesLASSO
# 14 = BayesCpi
# 15 = BayesB
# 16 = randomForest
# 17 = Trait Number
# 18 = Data Type (SNP/ DArT)
# 19 = Allele Seperator
# 20 =
# 21 =
# 22 =

#args <- c("D:/ICRISAT/Work/Analysis/Work/ISMU/ISMU", "SNP.csv", "Result.htm" ,1, 1, 1, -1, 2, "?", "Summary.csv", 1)
#args <- c("D:/ICRISAT/Work/Analysis/Work/ISMU/ISMU", "Genotype.csv", "Result.htm" ,1, 1, 1, -1, 2, "?", "Summary.csv", 1, "PhenoSum.html", "Phenotype.csv", 1)
#args <- c("Rscript", "dcn.R", "c:/RRes", "snp.csv", "sum.htm", 1, 1, 1, 2, ":", "-")

#rscript  D:\ICRISAT\Work\Analysis\Work\ISMU\ISMU\dcN.R "D:/ICRISAT/Work/Analysis/Work/ISMU/ISMU" "Genotype.csv" "Result.htm" 1 1 1 -1 2 "?" "Summary.csv" 2 "PhenoSum.html"



AWD       <- args[1]
AGDATA    <- args[2]
ARFILE    <- args[3]
AMISP  	  <- as.numeric(args[4])
APIC		  <- as.numeric(args[5])
AMAF		  <- as.numeric(args[6])
AMTYPE  <- as.numeric(args[7])
ASEP    <- args[8]
AMISSCH    <- args[9]
SumFileName <- args [10]
Engine  <- as.numeric(args[11])
#  Engine 
#  To use R , 1
#  To Use Fortran Alpha AGH 2
FilePhSUM <- args [12]
APDATA<- args [13]


ppi<-300  # Resolution
IR <- .5  # Image Ratio
SG <- 3   #Significant Decimal Digits

######################################################################################################################


### Generating random file names

GetRFileName <- function (start = "RES_")
{
  paste(start, substring(date(), 5, 7), substring(date(), 9, 10), "_", substring(date(), 12, 13), substring(date(), 15, 16), substring(date(), 18, 19), abs(round(rnorm(1),3)), sep = "")  
}


oldwd<-setwd(AWD)



#if()
####################################################################################
################### Prepare Phenotypic Summary File (FilePhSUM)#####################
####################################################################################
####################################################################################

BLUP <- read.csv(APDATA, na.strings=c(AMISSCH,"NA","NN", "*", "."), stringsAsFactors=F)
source(paste(oldwd, "/BasicStats.R", sep=""))

## Now make HTML Result File
GSRes <- file(FilePhSUM, "wt")   #Open in APPEND Mode to write results of Phenotype Summary

DCTitle <-"ISMU Genomic Selection Pipeline V.2.0"
cat(paste("<html> <head>  <title>", DCTitle,"</title>  <link rel=stylesheet href=ISMU.css type=text/css></head><body>"), file=GSRes)
close(GSRes) 



# Copy ISMU.CSS File to Current Working Directoy
FromCSSFile <- paste(oldwd, "/ISMU.css", sep="")
TOCSSFile <- paste(getwd(), "/ISMU.css", sep="")

file.copy(FromCSSFile, TOCSSFile)

# DONE, Creating HTML File

for ( i in 1:(ncol(BLUP)-1))
{
  # Create HTML File
  # As there are Problems with HTMLStart() in UNIX , we willl write our own code to create a HTML file
  TraitName<-colnames(BLUP)[1+i]
  x<-BLUP[,i+1]
  # Histogram for phenotypic trait with kernel density curve  
  PNGNamePlotHIST <- paste(getwd(), "/", GetRFileName(), ".png", sep="");
  THRes<-png(PNGNamePlotHIST, width=IR*10*ppi, height=IR*10*ppi, res=ppi)
  TRes<-hist(x, xlab=TraitName,main="Histogram",col="lightblue", prob=T)
  lines(density(x), lty=1,col="red",lwd=2)   
  TRes<-rug(x)
  TRes<-box()
  THRes<-dev.off()
  
  # Boxplot for phenotypic trait  
  PNGNamePlotBOX <- paste(getwd(), "/", GetRFileName(), ".png", sep="");
  THRes<-png(PNGNamePlotBOX, width=IR*10*ppi, height=IR*10*ppi, res=ppi)
  TRes<-boxplot(x, xlab=TraitName,main="Boxplot",col="lightblue",outcol="orange",outpch=19)
  THRes<-dev.off()
  
  
  # Descriptive Statistics and test for normality. Calling BasicStats.R
  THRes <- BasicStats(x)
  DescriptiveStat<-THRes$Basicstatistics
  NormTest <- THRes$Normality
  # print(DescriptiveStat)
  # print(NormTest)  
  
  
  # Write Trait heading in HTML File
  THRes<-HTML.title(paste("Phenotypic Summary for ", TraitName) , HR=1, file=FilePhSUM, Append=T)
  
  
  TRes<-HTML(DescriptiveStat, file=FilePhSUM, Append=T)
  # Add First Histogram
  TRes<-HTMLInsertGraph(PNGNamePlotHIST,file=FilePhSUM , WidthHTML="50%", append=T) 
  TRes<-HTML("<div align=\"center\">", file=FilePhSUM, Append=T)
  TRes<-HTML(paste("<b align=\"center\">Histogram for ", TraitName , "</b><br>"), HR=4, file=FilePhSUM, Append=T, Align="center")
  
  
  TRes<-HTML(paste("Test of Normality for ", TraitName), file=FilePhSUM, Append=T)
  TRes<-HTML(NormTest, file=FilePhSUM, Append=T)
  TRes<-HTML("<hr/>", file=FilePhSUM, Append=T)
  
  # Add Box Plot
  TRes<-HTMLInsertGraph(PNGNamePlotBOX, file=FilePhSUM , WidthHTML="50%", append=T) 
  TRes<-HTML("<div align=\"center\">", file=FilePhSUM, Append=T)
  TRes<-HTML(paste("<b align=\"center\">Box Plot for ", TraitName, "</b><br>"), HR=4, file=FilePhSUM, Append=T, Align="center")
  
  
  TRes<-HTML("<br/><br/><br/><br/>", file=FilePhSUM, Append=T)
  TRes<-HTML("<hr/>", file=FilePhSUM, Append=T)  
}
# Write Time and Logo
#TRes<-HTML("<hr/>", file=FilePhSUM, Append=T)
TRes<-HTML(paste("<p align=left>Results Generated by ISMU 2.0 : ", format(Sys.time(), "%a %b %d %X %Y")), file=FilePhSUM, Append=T)

# Close HTML File
GSRes <- file(FilePhSUM, "at")   #Open in APPEND Mode to write results of RR
cat("</body></html>", file=GSRes)
close(GSRes) 

####################################################################################
####################################################################################
####################################################################################
####################################################################################
####################################################################################
####################################################################################

setwd(oldwd)
sink()
file.remove(TFile)

